Seasteads housing fix (needs review!)

the-m4a opened the pull request:
@alimulhuda @alexeyOnGitHub 
This is ready for PR... for at least 3 reasons:
1. There is some significant shifting of code here, so the best thing to do may be just to do some playtesting.  Details on the shifting below.
2. There are minor issues to investigate, and I'd like eyes on that.
3. Patch drops sometime today and I expect we will need to see if there's fixes required.

**Code Work**
Previously, there were 3 copies of CityBannerManager.lua (basegame, exp1, exp2).  There were some significant differences between the basegame and expansion versions (the exp1 and exp2 versions were very similar).  Significant differences in this file between the basegame and expansion versions is also true for the unmodified version.

There are 5 new files, as the 3 CityBannerManager.lua files have been deleted:
- CityBannerManager_CQUI: Objects and functions common to the vanilla and expansion CityBannerManager files
- CityBannerManager_CQUI_basegame: Objects and functions specific to vanilla (basegame)
- CityBannerManager_CQUI_expansions: Objects and functions specific to Expansion1 and Expansion2
- CityBannerManager_CQUI_expansion1: Expansion1 wrapper file that simply does an _include("CityBannerManager_CQUI_expansions.lua")
- CityBannerManager_CQUI_expansion2: Expansion2 wrapper file that simply does an _include("CityBannerManager_CQUI_expansions.lua")

**Current Issues**
These issues are now fixed.  Huzzah!

~~I need to investigate further, but would appreciate help:~~
~~1. The settings are inconsistent in terms of being applied.  I don't know if this is a result of hotloading (making a change to the lua file and saving it will cause the game to reload it).  For example:~~
  ~~a. Disabling SmartBanners (district icons, etc) does not immediately take effect, but will once something forces the banners to refresh.~~
  ~~b. Enabling "show citizen icons" for mouseover doesn't do anything... it will never show the citizen icons (I know that it used to)~~
~~2. I only saw this in an old save, but there were district icons in the SmartBanner that were just a generic district icon, mixed in with the others.  I'm not sure if this is just due to an old save being loaded or what..~~

Thanks y'all.

the-m4a commented:
The icon weirdness I wrote about... maybe they're from Wonders?

![image](https://user-images.githubusercontent.com/8787640/85699163-25909680-b690-11ea-9262-b801996a95ec.png)

the-m4a commented:
Two other things before I forget:
1. The Maya housing for farms is effectively 1.5, not 1.0 as previously thought.  Unmodified game shows:
  - 1 Farm:  Housing From improvements is 1 (expected, as unmodified game drops the 0.5)
  - 2 Farms: Housing from improvements is 3 (would expect 2 if the value was 1.0, so it must be 1.5).
  - 3 Farms: Housing from improvements is 4..

2. I did test my changes in base, RF, and GS.
  - Verified the Maya housing calculation in base and GS, and tooltip shows 1.5 housing when creating a Farm
  - Verified Stepwell adds 1 housing before santiation, and 2 housing after (base game)
  - Verified banner functionality and whatnot in various game modes with various Civs

alexeyOnGitHub commented:
with the amount of changes, I think the most useful thing I can do is just test it locally as you suggested. any specific areas of the UI I should focus on?

alimulhuda commented:
Did you miss pushing a commit? I see you mention 'CityBannerManager_CQUI_expansion1' but I don't see that file and it's also not mentioned in the .modinfo file.

the-m4a commented:
@alimulhuda I believe you're right, I committed only locally... it was late :/.  It should be pushed things up now.

The extra icons in the smart banner were from wonders, so I updated the code to not show those.  I fixed up an issue I was seeing with the base game where some of the icons were offcenter -- likely because they were in the XML as "Image" instead of "Label" liek the others... 
(maybe they should all be image)?

Either way it the Vanilla way is messy and I'm not a fan, and would like to see if there's a way to match what it does in the expansions, but so far we haven't seen any new civs that come with unique districts that aren't already in vanilla

![image](https://user-images.githubusercontent.com/8787640/85810544-3e409100-b710-11ea-8e94-afee583ecfca.png)

the-m4a commented:
Figured out the issue with the settings not applying.   Simple typo, but the FileTuner wouldn't give me an error message unless I manually ran that code.  Frustrating, but happy to figure it out!

Infixo commented:
I loaded a save with those seasteads that triggered an issue.
a) All banners are gone. But I accidentaly found them - all stacked up on one of the cities.
b) Seasteads are counted propely at least in the city panel (I can open it via BRS).
c) There are no errors in the lua.log, just tons of debug messages.

Infixo commented:
It seems that the missing banners happen only for previously started games. The new games don't have this issue.

Infixo commented on Assets/UI/WorldView/citybannermanager_CQUI_expansions.lua:
```diff
@@ -0,0 +1,188 @@
+function Initialize()
+    print_debug("CityBannerManager_CQUI_Expansions: Initialize CQUI CityBannerManager")
+    -- Events are initialized in the common file
+    LateInitialize();
```
You should not call LateInitialize() from Initialize().

the-m4a replied:
I need to rename lateinitialize, I think.  If I remember correct I modified citybannermanager does not declare its own LateInitialize to override, and it’s unclear to me how often one can call Initialize... this is an attempt to not have 3 different Initialize calls for the various CityBannerManager files
(unmodifed, citibannermanager_cqui, citybannermanager_cqui_(basegame/expansions))

Infixo replied:
It does, it uses a different name. It is always hooked to ContextPtr:SetInitHandler(), and here it is OnContextInitialize. But there is no need to override that.

Infixo commented on Assets/UI/WorldView/citybannermanager_CQUI_expansions.lua:
```diff
@@ -0,0 +1,188 @@
+                    local isBuilt = pCityDistricts:HasDistrict(districtInfo.Index, true);
+                    if (isBuilt == true
+                        and districtInfo.Index ~= 0
+                        and districtInfo.DistrictType ~= "DISTRICT_GOVERNMENT"
```
Hm? Why is Gov Plaza excluded? It is a pretty important district.

the-m4a replied:
I think I had it confused for the city center, good catch will fix

Infixo replied:
Fixed.

Infixo commented on Assets/UI/CQUICommon.lua:
```diff
@@ -7,33 +7,41 @@
 -- ===========================================================================
 --CQUI setting control support functions
 -- ===========================================================================
 function print_debug(str)
```
This is so not flexible - can't pass multiple args. Please use multi args (print_debug(...) and then inside print(...)).

the-m4a commented:
Thanks for taking a look! Much appreciated.
I saw that issue once, but it was inconsistent, as I could not get it to happen again.  So, good kinda to see that someone else observed it as it rules out weirdness i might of created during the process of working on this.  I likely won’t be able to take a look until later tonight unfortunately.

I did not see this until this most recent set of changes where I made all the variables local to just citybannermanager_cqui.lua.  I don’t think that should make a diff but it somehow appears to make a difference.

I saw it on a maya save game, vanilla rule set (no RF or GS enabled).  I loaded a later version of that same Maya game and it was fine.  India save game, vanilla, it was fine.  All the GS saves I tried, also fine.  And same - nothing in the lua log.  I need to go back and try to load those games as the first thing loaded after starting Civ, maybe it’s due to something that gets cached after one load game occurs?

I do need to revert the debug print back to off before the PR is completed as yes, it is noisy at the moment.

Infixo commented:
The disappearing banners are irregular. Sometimes the same savefile loads ok, and sometimes not ok. The worst kind of issue to happen. I am trying to find out what the heck is going on.

the-m4a commented:
Awesome thank you.  I wish there was a better way to debug these things rather than putting print statements everywhere.

Infixo commented:
Am I allowed to make changes to the code directly? I mean can I make a commit to the branch?

the-m4a commented:
I believe unless alexey reverted your permissions you should be able to push changes to the same branch, yes

Infixo commented:
I think I found out the cause and know how to fix it.

the-m4a commented:
What was it? Looking at the changes on my phone and it’s not making the change obvious

Infixo commented:
The disappearing banners were caused by this line:
LuaEvents.CQUI_SettingsInitialized.Add( CQUI_OnSettingsUpdate );

It caused a Reload() to be called probably before the context was fully initialized and thus producing weird results. I suppose the irregularity was coming from the fact that tables are not ordered in Lua and probably contexts are not initialized in an undetermined order. So, if banners were initialized before SettingInitialized then all was ok, if not then some weird results happen.

My comment about LateInitialize() was basically about that (not specifically the issue, but the general rule) and it turned out it was spot on.

Infixo commented:
Found and fixed 2 more bugs.
1. Mouse over was not showing the next tile.
2. Debug was always on even if set to false.

Infixo commented:
Re: debug flag.
In Lua, a number is considered true if evaluated for boolean. Only nil is considered false. So, if you just assign 0 to a boolean (like it was in the code), then it will be true.

the-m4a commented:
@alexeyOnGitHub @alimulhuda ? y'all around to review?

alimulhuda commented:
Sorry haven't had a chance to properly look over the changes, the weekend was a lot more busy than usual. July 1 is a day off here, I should hopefully have a chance to look over stuff then.

Infixo commented:
I suggest to change the rule to "1 review required". If someone wants to contribute, then he can review or contribute at any time. But the 2-reviews-rule is stalling the project. People are busy with IRL, the team is just too small, and issues are piling up.

alexeyOnGitHub commented:
I don't mind changing from "2 reviews" rule to "1 review", if the rest of the team is ok with this. 
2 reviews were designed to make the dev process safer in the beginning when we are just getting people to work on this repo.

the-m4a commented:
I am also okay with requiring 1 reviewer.  I will also attempt to try and keep the changes smaller - though for this re-work stuff, that's tough to do.

For the "change the 2-spaces to 4-spaces" change, there will be a number of files, but the only differences will be in the blanks.  GitHub isn't great for those comparisons though.

(There's at least 1 file where I added parenthesis to the if statements, which I want to undo before submitting that PR.  The 4 spaces (or tabs at 4 spaces) really improves the readability)

I also wonder if it's worthwhile to have a "pre-production environment" version of the mod that people can subscribe to and help test out.  Seems it may be easier than attracting them here to do PRs and manually install...

alexeyOnGitHub commented:
ok, I changed "2 reviews" to "1 review" in the repo settings.

I like the idea of having a staging env. not sure if Steam Workshop supports this concept natively. if it does not, then one option could be to have a separate mod ID and a descriptive name like `blah-blah...-DEV-SNAPSHOT` or something. then people can subscribe to one or the other in Steam workshop. I would really want this release process to be automated though, so that there is a lower chance of a human error. the way I see it, "dev release" script could hack the mod ID plus mod name on local disk, then zip and upload the whole thing to Steam Workshop. this can be set up in a continuous integration.

Infixo commented:
Re: 2 vs. 4 spaces
I am not gonna stop anybody if you want to go ahead and reformat the files.
However, I think it is just easier and more productive to do those changes when the file is to be modified because of some other reasons / incidents / etc.

alimulhuda commented:
I also think the the whole nightly build idea is worth doing. In my mind the idea you all were talking about is loosening the restrictions for merging to master and making it so that the master branch gets uploaded to Steam every time there is a change (or every night). Then we just need to think of a way to decide when to promote a nightly build to stable (anything other than a simple promotion from nightly to stable might be too involved). That way we can have people easily try out both the bleeding edge and the "should work without major bugs" versions.

the-m4a commented:
Thanks y'all.

I just got done with work (stupid long projects and SQL making my face hurt), so I won't send the PR for the spaces thing tonight.  No need to wait on it either if you have changes you want to send up, I can always incorporate changes into my branch later...

I figure just will get them done, so the files stay consistent when we're making the other changes.  Changing the spacing in the files while also making fixes would make it tougher to pick out what's changed.

We'd have to script some sort of thing to do that updating, did we ever figure if it was actually possible?

alimulhuda commented on Assets/UI/WorldView/citybannermanager_CQUI_expansions.lua:
```diff
@@ -0,0 +1,183 @@
+-- ===========================================================================
+-- Cached Base Functions (Expansions only)
+-- ===========================================================================
+BASE_CQUI_CityBanner_Initialize = CityBanner.Initialize;
+BASE_CQUI_CityBanner_Uninitialize = CityBanner.Uninitialize;
+BASE_CQUI_CityBanner_UpdateInfo = CityBanner.UpdateInfo;
+BASE_CQUI_CityBanner_UpdatePopulation = CityBanner.UpdatePopulation;
+BASE_CQUI_CityBanner_UpdateRangeStrike = CityBanner.UpdateRangeStrike;
+BASE_CQUI_CityBanner_UpdateStats = CityBanner.UpdateStats;
```
The unmodded game has CityBanner:Initialize and not CityBanner.Initialize, for the expansions. In Lua are they (: vs .) interchangeable?

alimulhuda replied:
Okay figured out the difference, never mind.

alimulhuda commented on CQUI.modinfo:
```diff
@@ -1021,7 +1082,9 @@
     <File>Assets/UI/ToolTips/plottooltip_CQUI.lua</File>
     <File>Assets/UI/Utilities/extendedrelationship.lua</File>
-    <File>Assets/UI/WorldView/citybannermanager.lua</File>
+    <!-- <File>Assets/UI/WorldView/citybannermanager.lua</File> -->
+    <File>Assets/UI/WorldView/citybannermanager_CQUI_basegame.lua</File>
```
Minor cleanup, comment leftover.
